// views/m/common/CitySelector/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as Vue from 'vue';
import * as city from 'city';
import * as autocomplete from 'autocomplete';

class VPresenter extends project.VPresenter {
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._watchEvent();
        var that = this;
        var a = autocomplete;//hack

        var v_currCity = this.find('#current')[0];
        var curr_city = model.globalData.city;
        new Vue({
            el: v_currCity,
            data: curr_city
        })
        //渲染列表
        var v_el = this.find('.city-all')[0];
        var city_list = city;
        new Vue({
            el: v_el,
            data: { city_list: city_list },
            methods: {
                chooseCity: function (event) {
                    //选择城市，更改全局数据，关闭弹窗
                    if (event.target.className == 'city-item') {
                        console.log(event.target.innerHTML);
                        var sel = {
                            city_id: event.target.getAttribute('city_id'),
                            city_name: event.target.innerHTML
                        }
                        model.globalData.selected_city.city_id = sel.city_id;
                        model.globalData.selected_city.city_name = sel.city_name;
                        funs.viewHistory.go(-1);
                    }
                }
            }
        });
        //热门城市
        var source: any[] = []
        var hot_city: any[] = [];
        for (var key in city) {
            var value = city[key];
            for (var i = 0; i < value.length; i++) {
                source.push({ label: value[i].city_name, value: value[i].city_id });
                if (value[i].is_common == '1') {
                    hot_city.push(value[i]);
                }
            }
        };
        var v_hot = this.find('#hots')[0];
        new Vue({
            el: v_hot,
            data: {
                hot_city: hot_city
            },
            methods: {
                chooseCity: function (event) {
                    //选择城市，更改全局数据，关闭弹窗
                    if (event.target.className == 'city-button') {
                        console.log(event.target.innerHTML);
                        var sel = {
                            city_id: event.target.getAttribute('city_id'),
                            city_name: event.target.innerHTML
                        }
                        model.globalData.selected_city.city_id = sel.city_id;
                        model.globalData.selected_city.city_name = sel.city_name;
                        funs.viewHistory.go(-1);
                    }
                }
            }
        });
        //搜索城市自动完成提示
        this.find('.input-div').autocomplete({
            source: source,
            //选中事件
            select: function (event, ui) {
                $(this).val(ui.item.label);
                var sel = {
                    city_id: ui.item.value,
                    city_name: ui.item.label
                }
                model.globalData.selected_city.city_id = sel.city_id;
                model.globalData.selected_city.city_name = sel.city_name;
                event.preventDefault();
                funs.viewHistory.go(-1);
            }
        }).autocomplete('instance')._renderItem = function (ul, item) {
            return $("<li><a href='#'>" + item.label + "</a></li>").appendTo(ul);
        }
    }
}
export = VPresenter;
