// views/global/common/Funs/index.ts
import * as tomato from '@po-to/tomato';

export let viewHistory:tomato.ViewHistory = new tomato.ViewHistory(3,4);

let metaDataSites: { [key: string]: string } = (`{SITES}` as any);

let siteMap = (function (sites) {
    let map = {};
    for (let domain in sites) {
        let folder = sites[domain];
        if (folder) {
            map[folder] = domain + ":`{SITEPORT}`";
        }
    }
    return map;
})(metaDataSites)

export function moduleToUrl(id: string): string {
    for (let folder in siteMap) {
        if (id.startsWith(folder)) {
            return "`{HTTP}`" + siteMap[folder] + id.replace(folder, '') + '/?__page__=&__request__=amd';
        }
    }
    return '';
}

export function setCookie(key: string, value: string, expiredays?: number) {
    let api = "`{APIURL}`";
    let apiHost = api.substr(api.indexOf('//')+2);
    let cookie: string[] = [apiHost+'$'+key + "=" + encodeURIComponent(value)];
    if (expiredays) {
        let exdate = new Date();
        exdate.setDate(exdate.getDate() + expiredays);
        cookie.push("expires=" + (exdate as any).toGMTString());
    }
    cookie.push('path=/');
    let arr = document.domain.split(".");
    arr = ['',arr[arr.length-2],arr[arr.length-1]];
    cookie.push('domain='+arr.join("."));
    document.cookie = cookie.join("; ")
}
export function setSession(key: string, value: string) {
    localStorage.setItem(key, value);
}
export function getSession(key: string) {
    localStorage.getItem(key);
}
export function loadPage(mod:tomato.VPresenter){
    let vpid = mod.vpid.replace(/&?\__page__=[^&]*/,'').replace(/&?\__request__=[^&]*/,'').replace(/\/?\?$/,'');
    window.history.pushState("","",vpid);

    let cmd = new tomato.Cmd();
    let curMod = tomato.application.content;
    cmd.execute = function(){
        (mod.view as JQuery).attr('data-back','');
        tomato.application.appendChild(mod);
        let vpid = mod.vpid.replace(/&?\__page__=[^&]*/,'').replace(/&?\__request__=[^&]*/,'').replace(/\/?\?$/,'');
        window.history.pushState("","",vpid);
        this.success();
    }
    cmd.undo = function(){
        curMod && (curMod.view as JQuery).attr('data-back','1');
        tomato.application.appendChild(curMod as tomato.VPresenter);
        this.success();
    }
    cmd.redo = function(){
        (mod.view as JQuery).attr('data-back','');
        tomato.application.appendChild(mod);
        this.success();
    }
    viewHistory.uriPush(cmd);
}
