// views/m/project/EditCurrentItem/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as funs from "views/global/common/Funs";
import ChatRoom = require('views/m/ChatRoom');
import ImageUploader = require('views/m/project/ImageUploader');

class VPresenter extends project.VPresenter {
    /**模块 */
    private EditCurrentItem_vue: any;
    /**循环选择的数据源 */
    private demand_config;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);

        var vueDom = this.find("#EditCurrentItem_vue").get(0);
        var demand_config = this.demand_config = model.globalData.demand_config;
        var user = model.globalData.user;
        var demand = model.globalData.demand;
        var that = this;
        let EditCurrentItem_vue: any = this.EditCurrentItem_vue = new Vue({
            el: vueDom,
            data: {
                typeClick: false,
                locationClick: false,
                colorClick: false,
                spanClick: false,
                OtherClick: false,
                otherSpan: '',
                clicked: true,
                user: user,
                demand: demand,
                Url: {
                    typeUrl: [demand_config.typeOptions[2], demand_config.typeOptions[5], demand_config.typeOptions[3], demand_config.typeOptions[1], demand_config.typeOptions[4], demand_config.typeOptions[0],],
                    locationUrl: [demand_config.locationOptions[1], demand_config.locationOptions[3], demand_config.locationOptions[2]],
                    colorUrl: [demand_config.colorOptions[3], demand_config.colorOptions[2], demand_config.colorOptions[1]],
                    spanUrl: [demand_config.spanOptions[1], demand_config.spanOptions[2], demand_config.spanOptions[3], demand_config.spanOptions[4]],
                }
            },
            watch: {
                'demand.address': "checkInfo",
                'demand.size': "checkData",
                'demand.budget': "checkData"
            },
            methods: {
                checkData: function () {
                    //  EditCurrentItem_vue.demand.budget = EditCurrentItem_vue.demand.budget.replace(/[^\d]/g, '');
                    EditCurrentItem_vue.demand.budget = EditCurrentItem_vue.demand.budget.replace(/[^\d]/g, '');
                    EditCurrentItem_vue.demand.size = EditCurrentItem_vue.demand.size.replace(/[^\d\.]/g, '');
                    EditCurrentItem_vue.checkInfo();
                },
                checkInfo: function () {
                    //检验必填项
                    if (EditCurrentItem_vue.demand.address && EditCurrentItem_vue.demand.size) {
                        EditCurrentItem_vue.clicked = false;
                    } else {
                        EditCurrentItem_vue.clicked = true;
                    }
                },
                //展开下拉菜单
                showType: function () {
                    EditCurrentItem_vue.typeClick = !EditCurrentItem_vue.typeClick;
                    EditCurrentItem_vue.locationClick = false;
                    EditCurrentItem_vue.colorClick = false;
                    EditCurrentItem_vue.spanClick = false;
                },
                showLocation: function () {
                    EditCurrentItem_vue.locationClick = !EditCurrentItem_vue.locationClick;
                    EditCurrentItem_vue.typeClick = false;
                    EditCurrentItem_vue.colorClick = false;
                    EditCurrentItem_vue.spanClick = false;
                },
                showColor: function () {
                    EditCurrentItem_vue.colorClick = !EditCurrentItem_vue.colorClick;
                    EditCurrentItem_vue.typeClick = false;
                    EditCurrentItem_vue.locationClick = false;
                    EditCurrentItem_vue.spanClick = false;
                },
                showSpan: function () {
                    if (EditCurrentItem_vue.OtherClick) {
                        EditCurrentItem_vue.showCloseAll();
                    } else {
                        EditCurrentItem_vue.spanClick = !EditCurrentItem_vue.spanClick;
                        EditCurrentItem_vue.typeClick = false;
                        EditCurrentItem_vue.locationClick = false;
                        EditCurrentItem_vue.colorClick = false;
                    }

                },
                showCloseAll: function () {
                    EditCurrentItem_vue.spanClick = false;
                    EditCurrentItem_vue.typeClick = false;
                    EditCurrentItem_vue.locationClick = false;
                    EditCurrentItem_vue.colorClick = false;
                    EditCurrentItem_vue.OtherClick = false;
                },
                //参数选择
                getType: function (event) {
                    let id = event.target.getAttribute('value');
                    let name = event.target.innerHTML;
                    that.setInit(id);

                    EditCurrentItem_vue.showType();

                },
                getLocation: function (event) {
                    EditCurrentItem_vue.demand.location = event.target.getAttribute('value');
                    EditCurrentItem_vue.demand.location_name = event.target.innerHTML;
                    EditCurrentItem_vue.showLocation();
                },
                getColor: function (event) {
                    EditCurrentItem_vue.demand.color = event.target.getAttribute('value');
                    EditCurrentItem_vue.demand.color_name = event.target.innerHTML;
                    EditCurrentItem_vue.showColor();
                },
                getSpan: function (event) {
                    let id = event.target.getAttribute('value');
                    let name = event.target.innerHTML;
                    // EditCurrentItem_vue.demand.span = name;
                    EditCurrentItem_vue.demand.span_id = id;
                    if (4 == id) {
                        EditCurrentItem_vue.showCloseAll();
                        EditCurrentItem_vue.OtherClick = true;
                        EditCurrentItem_vue.otherSpan = '';
                    } else {
                        EditCurrentItem_vue.demand.span = name;
                        EditCurrentItem_vue.showSpan();
                    }
                },
                getSure: function () {
                    EditCurrentItem_vue.demand.span = EditCurrentItem_vue.otherSpan;
                    EditCurrentItem_vue.showCloseAll();
                },
                getClean: function () {
                    // EditCurrentItem_vue.otherSpan = '';
                    EditCurrentItem_vue.showCloseAll();
                    setTimeout(function () {
                        EditCurrentItem_vue.showSpan();
                    }, 0);

                }, /**提交 */
                submitEdit: function () {
                    var demand_id = parseInt(EditCurrentItem_vue.demand.id);
                    var type = parseInt(EditCurrentItem_vue.demand.type);
                    var size = parseFloat(EditCurrentItem_vue.demand.size);
                    var color = parseInt(EditCurrentItem_vue.demand.color);
                    var location = parseInt(EditCurrentItem_vue.demand.location);
                    var span = EditCurrentItem_vue.demand.span;
                    var budget = parseInt(EditCurrentItem_vue.demand.budget);
                    var address = EditCurrentItem_vue.demand.address;
                    var image = EditCurrentItem_vue.demand.image;
                    var note = EditCurrentItem_vue.demand.note;
                    //提交图片
                    tomato.getVPresenter<ImageUploader>(funs.moduleToUrl('m/project/ImageUploader'), (ImageUploader) => {
                                ImageUploader.fileUpload();
                            })
                    api.EditDemand({ demand_id, type, size, color, location, span, budget, address,note }, function (data: any) {
                        // window.location.reload();
                        funs.viewHistory.go(-1);
                        // 返回详情时调用预算金额计算方法
                        tomato.getVPresenter<ChatRoom>(funs.moduleToUrl('m/ChatRoom'), (ChatRoom) => {
                            ChatRoom.setBudget();
                        });

                    }, function (error) {
                        console.log(error);
                        alert("提交失败！");
                    });
                },
            }
        });
        this._watchEvent();
    }
    /**编辑项目时，将demand的type传过来，以便做样式上的初始化 */
    public setInit(typeId) {
        let def = this.demand_config.typeOptions[typeId].def;
        let name = this.demand_config.typeOptions[typeId].name;

        this.EditCurrentItem_vue.demand.type = typeId;
        this.EditCurrentItem_vue.demand.type_name = name;

        this.EditCurrentItem_vue.Url.locationUrl = this.demand_config.locationOptions = def.locationOptions;
        this.EditCurrentItem_vue.Url.colorUrl = this.demand_config.colorOptions = def.colorOptions;
        this.EditCurrentItem_vue.Url.spanUrl = this.demand_config.spanOptions = def.spanOptions;


        this.EditCurrentItem_vue.demand.location = def.location;
        this.EditCurrentItem_vue.demand.location_name = this.EditCurrentItem_vue.Url.locationUrl[def.location].name;
        this.EditCurrentItem_vue.demand.color = def.color;
        this.EditCurrentItem_vue.demand.color_name = this.EditCurrentItem_vue.Url.colorUrl[def.color].name;
        this.EditCurrentItem_vue.demand.span = this.EditCurrentItem_vue.Url.spanUrl[def.span].name;
        this.EditCurrentItem_vue.demand.span_id = def.span;

        //检测必填项
        this.EditCurrentItem_vue.checkInfo();
    }
}
export = VPresenter;
