// views/m/Index/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";


class VPresenter extends project.VPresenter{
    private _header:tomato.VPresenter;
     private footer: tomato.VPresenter;
     private inquiryToTop:number;
     private footerH:number;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?:string){
        super(view,parent,vpid);
        this._watchEvent();
        
    }
     /** */
    public init() {
        return tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/project/InquiryLayer'), (InquiryLayer) => { this.footer = InquiryLayer; })
    }
    public getFooter(){
        if(model.globalData.demand.id && model.globalData.demand.id != '0'){
            return null
        }else{
            if(!this.footerH){
                this.footerH = $(this.footer.view).height();
            }
            return this.footer.view;
        }
    }
    private _showFooter(){
        var body = this.view[0].parentNode;
        if ($(body).scrollTop() >= this.inquiryToTop+100){
            if($(this.footer.view).height() == 0){
                $(this.footer.view).height(this.footerH);
                (this.parent as tomato.Dialog).refreshLayout();
            }
        }else{
            if($(this.footer.view).height() != 0){
                $(this.footer.view).height(0);
                (this.parent as tomato.Dialog).refreshLayout();
            }
        }
    }
    protected _afterInstallTo(){
        //如果登陆且有项目，则返回
        if(model.globalData.demand.id && model.globalData.demand.id != '0'){
            return
        }
        //滚动监听，出现底部询价条
        if(!this.inquiryToTop){
            this.inquiryToTop = this.find('#Inquiry').offset().top;
        }
        $(this.view[0].parentNode).scroll(()=>{
            this._showFooter();
        })
        this._showFooter();
    }
    protected _afterUninstallTo(){
        $(this.view[0].parentNode).off('scroll');
    }
}
export = VPresenter;
