// views/m/common/Menu/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as Vue from 'vue';
import * as api from 'views/global/common/API'

class VPresenter extends project.VPresenter{
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?:string){
        super(view,parent,vpid);
        
        //this._els = this._getElements();3
        var el = this.find('.v-user-info')[0];
        var user_info = model.globalData.user;
        new Vue({
            el: el,
            data: user_info,
            methods:{
                 logIn:function(){
                     tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/user/LoginForm'), (LoginForm) => {
                            //登录窗
                            tdom.open(LoginForm, tdom.DialogTarget.Blank, {
                                size: { width: '90%', height: tomato.DialogSize.Content },
                                effect: tdom.DialogEffect.scale,
                                masked: true
                            });
                        })
                 },
                 logOut:function(){
                        api.Logout(function(){
                            //success
                            window.location.reload();
                        },function(){
                            //fail
                            alert('注销登录失败，请重试')
                        })
                 }
            }
        });
        this._watchEvent();
    }
    _evt_load(url:string){
        (this.parent as tomato.Dialog).close();
        super._evt_load(url);
    }
}
export = VPresenter;
