// views/m/project/InquiryLayer/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import * as tdom from "@po-to/tomato-jquery";
import InquiryEdit = require('views/m/project/InquiryEdit');
import ChatRoom = require('views/m/ChatRoom');

class VPresenter extends project.VPresenter {
    /**模块 */
    public InquiryLayer_vue;
    private _layerBodyDialog: tomato.Dialog | undefined;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        var vueDom = this.find("#InquiryLayer_vue").get(0);
        var city = model.globalData.selected_city;
        var user = model.globalData.user;
        var demand = model.globalData.demand;
        var that = this;
        let InquiryLayer_vue: any = this.InquiryLayer_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                city: city,
                demand: demand,
                typeClick: false,
                typeName: '户外广告屏'
            },
            watch: {
                'code': "checkData",
            },
            methods: {
                /**选择城市 */
                openCitySel: function (event) {
                    tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
                        funs.loadPage(CitySelector);
                    })
                },
                /**选择类型 */
                showType: function () {
                    //展开选择项
                    if (that._layerBodyDialog) {
                        that._layerBodyDialog.close();
                        InquiryLayer_vue.typeClick = false;
                    } else {
                        InquiryLayer_vue.typeClick = true;
                        console.log(InquiryLayer_vue.typeClick)
                        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/project/InquiryLayerBody'), (InquiryLayerBody) => {
                            that._layerBodyDialog = tdom.open(InquiryLayerBody, tdom.DialogTarget.Blank, {
                                // className: "Dialog-Menu",
                                effect: tdom.DialogEffect.slideUp,
                                masked: true,
                                size: { width: tomato.DialogSize.Full, height: tomato.DialogSize.Content },
                                 position:{
                                    x:0,
                                    y:tomato.DialogPosition.Bottom
                                },
                                offset: { x: 0, y: [that.view,"*-1"] }
                            });
                            that._layerBodyDialog['_afterClose'] = () => {
                                that._layerBodyDialog = undefined;
                            }
                        })
                    }

                },
                /**询价 */
                inquiryEvent: function () {
                    tomato.getVPresenter<InquiryEdit>(funs.moduleToUrl('m/project/InquiryEdit'), (InquiryEdit) => {
                        InquiryEdit.setProjeType(demand.type);
                        console.log(demand.type);
                        funs.loadPage(InquiryEdit);
                    })
                },
                /**打开洽谈室 */
                openChatRoom: function (event) {
                    //将临时数据更改为自己的项目的数据
                    model.globalData.current_demand = model.globalData.demand;
                    tomato.getVPresenter<ChatRoom>(funs.moduleToUrl('m/ChatRoom'), (ChatRoom) => {
                        //弹出详情页
                        ChatRoom.openChatRoom();
                        tomato.application.appendChild(ChatRoom);
                    })
                }
            }
        });

        //this._watchEvent();
        //this._els = this._getElements();
    }
}
export = VPresenter;
