// views/m/common/Header/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as api from "views/global/common/API";


class VPresenter extends project.VPresenter {
    private _menuDialog: tomato.Dialog | undefined;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._watchEvent();
        this._els = this._getElements();
        // (this.find('.collapse') as any).collapse('hide');
        // this.find('#collapseOne').on('show.bs.collapse', function () {
        //     $('body').append("<div class='menu-mask'></div>");
        // }).on('hidden.bs.collapse', function () {
        //     $('.menu-mask').remove();
        // });

    }
    _evt_showMenu(data: any, target: HTMLElement, hit: HTMLElement) {
        if (this._menuDialog) {
            this._menuDialog.close();
        } else {
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/Menu'), (Menu) => {
                this._menuDialog = tdom.open(Menu, tdom.DialogTarget.Blank, {
                    className: "Dialog-Menu",
                    effect: tdom.DialogEffect.slideDown,
                    masked: true,
                    size: { width: tomato.DialogSize.Full, height: tomato.DialogSize.Content },
                    position: { x: 0, y: 0 },
                    offset: { x: 0, y: $(this.view[0]) }
                });
                this._menuDialog['_afterClose'] = () => {
                    this._menuDialog = undefined;
                }
            })
        }
    }
}
export = VPresenter;
