// views/m/user/Code/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as funs from "views/global/common/Funs";

class VPresenter extends project.VPresenter {
    /**模块 */
    private Code_vue: any;
    /**手机号 */
    private phone;
    /**demand_id */
    private demand_id;
    /**验证码倒计时 */
    private codeGain;
    /**span */
    private codeParent;

    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);

        var vueDom = this.find("#Code_vue").get(0);
        var user = model.globalData.user;
        var demand = model.globalData.demand;
        let Code_vue: any = this.Code_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                demand: demand,
                data: '59s后重新获取验证码',
                code: '',
            },
            watch: {
                'code': "checkData",
            },
            methods: {
                checkData: function () {
                    Code_vue.code = Code_vue.code.replace(/[^\d]/g, '');
                }
            }
        });

        this.phone = Code_vue.user.phone;
        // this.demand_id = Code_vue.demand.demand_id;
        this.codeGain = this.find(".c-code-gain");
        this.codeParent = this.find(".c-code-put");
        this.getCode();
        this.inputFocus();

        this._watchEvent();
        this.addListener(tomato.DialogEvent.Closed,function(){
            funs.viewHistory.go(-1);
        })
    }
    setDemandId(demand_id){
        this.demand_id = demand_id;
    }

    private getCode = function () {
        var that = this;
        that.codeCountDown();
    }

    private codeCountDown = function () {
        var that = this;
        that.codeGain.html('60s后重新获取验证码').prop('disabled', true);
        var wait = 60;
        var timer = setInterval(function () {
            wait--;
            that.codeGain.html(wait + 's后重新获取验证码');
            if (wait < 1) {
                clearInterval(timer);
                this.timer = null;
                that.codeGain.html('重新获取验证码').prop('disabled', false).addClass('code-blue');
            }
        }, 1000)
    }

    private _evt_getCode = function () {
        var that = this;
        var phone = that.phone;
        //a02发送手机验证码接口
        api.SendSmsCaptcha(phone, function (json) {
            //倒计时
            that.codeCountDown();
        }, function (error) {
            alert(error.message);
        });
    }

    private inputFocus = function () {
        var that = this;

        this.codeParent.bind('input propertychange', 'input', function (event) {
            var sum = this.innerHTML;
            if (4 == sum.length) {
                $(this).attr('contenteditable','false');
                console.log(sum + '4位数字，尝试校验...');
                that.verifyPasscode(sum);
            }
        });
    }

    private verifyPasscode = function (passcode) {
        var that = this;
        var data = {
            phone:that.phone,
            demand_id: that.demand_id,
            sms_captcha:passcode
        }
        api.ActivateDemand(data,function(){
            console.log("创建成功");
            localStorage.setItem("chatroom", 'open');//存储变量名为key，值为value的变量
            window.location.reload();
        },function(data:tomato.PError){
            alert(data.note);
            that.codeParent.html('').attr('contenteditable','true');;
        })
        // // c33-验证激活待完善项目
        // a_demand_activate_project(this.phone, passcode, demand_id, {
        //     succ: function (json) {


        //         // 请求成功，则记一个cookie，以便刷新页面时如果有这个cookie，则打开聊天室
        //         // 而打开弹层的判断要在CustomerIterPage.js中操作

        //         localStorage.setItem("chatroom", 'open');//存储变量名为key，值为value的变量
        //         window.location.reload();

        //     },
        //     fail: function (json) {
        //         alert(json.msg)
        //         $(that._els.incode).find('input').val('');
        //         $(that._els.incode).eq(0).find('input').focus();

        //     }
        // });
    }
}
export = VPresenter;
