// views/m/project/InquiryEdit/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as tdom from "@po-to/tomato-jquery";
import * as funs from "views/global/common/Funs";
import LoginForm = require('views/m/user/LoginForm');
import chat = require('views/m/ChatRoom');

class VPresenter extends project.VPresenter {
    /**模块 */
    private InquiryEdit_vue: any;

    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        var vueDom = this.find("#InquiryEdit_vue").get(0);
        var demand_config = model.globalData.demand_config;
        var user = model.globalData.user;
        var selected_city = model.globalData.selected_city;
        var demand = model.globalData.demand;
        let InquiryEdit_vue: any = this.InquiryEdit_vue = new Vue({
            el: vueDom,
            data: {
                picked: '',
                typePicked: '2',
                locationPicked: '2',
                colorPicked: '2',
                spanPicked: '2',
                totalShow: true,
                clicked: true,
                showError: false,
                checkPhoneData: '',
                user: user,
                demand: demand,
                selected_city: selected_city,
                Url: {
                    typeUrl: [demand_config.typeOptions[2], demand_config.typeOptions[5], demand_config.typeOptions[3], demand_config.typeOptions[1], demand_config.typeOptions[4], demand_config.typeOptions[0],],
                    locationUrl: [demand_config.locationOptions[1], demand_config.locationOptions[3], demand_config.locationOptions[2]],
                    colorUrl: [demand_config.colorOptions[3], demand_config.colorOptions[2], demand_config.colorOptions[1]],
                    spanUrl: [demand_config.spanOptions[1], demand_config.spanOptions[2], demand_config.spanOptions[3], demand_config.spanOptions[4]],
                }
            },
            watch: {
                typePicked: "chooseType",
                locationPicked: "chooseLocation",
                colorPicked: "chooseColor",
                spanPicked: "chooseSpan",
                // 'user.phone': "checkData",
                'user.nick': "checkData",
                'demand.size': "checkData",
            },
            methods: {
                chooseType: function () {
                    let index = InquiryEdit_vue.typePicked;
                    let def = demand_config.typeOptions[index].def;
                    InquiryEdit_vue.demand.type = demand_config.typeOptions[index].id;
                    InquiryEdit_vue.demand.type_name = demand_config.typeOptions[index].name;

                    InquiryEdit_vue.locationPicked = def.location;
                    InquiryEdit_vue.colorPicked = def.color;
                    InquiryEdit_vue.spanPicked = def.span;

                    InquiryEdit_vue.Url.locationUrl = demand_config.locationOptions = def.locationOptions;
                    InquiryEdit_vue.Url.colorUrl = demand_config.colorOptions = def.colorOptions;
                    InquiryEdit_vue.Url.spanUrl = demand_config.spanOptions = def.spanOptions;

                    InquiryEdit_vue.chooseLocation();
                    InquiryEdit_vue.chooseColor();
                    InquiryEdit_vue.chooseSpan();
                },
                chooseLocation: function () {
                    let index = InquiryEdit_vue.locationPicked;

                    InquiryEdit_vue.demand.location = demand_config.locationOptions[index].id;
                    InquiryEdit_vue.demand.location_name = demand_config.locationOptions[index].name;
                },   
                chooseColor: function () {
                    let index = InquiryEdit_vue.colorPicked;

                    InquiryEdit_vue.demand.color = demand_config.colorOptions[index].id;
                    InquiryEdit_vue.demand.color_name = demand_config.colorOptions[index].name;

                }, 
                chooseSpan: function () {
                    let index = InquiryEdit_vue.colorPicked;

                    InquiryEdit_vue.demand.span_id = demand_config.spanOptions[index].id;
                    InquiryEdit_vue.demand.span = demand_config.spanOptions[index].name;
                },
                //展开选择列表
                showList: function () {
                    $(view).find('.inquiry-total').addClass('hide');
                    $(view).find('.inquiry-list').removeClass('hide');
                    $(view).css('height','9.2rem').parent().css({'overflow-y':'scroll'});
                },
                //限制输入
                checkData: function () {
                    InquiryEdit_vue.user.phone = InquiryEdit_vue.user.phone.replace(/[^\d]/g, '');
                    InquiryEdit_vue.user.nick = InquiryEdit_vue.user.nick.replace(/[^(a-zA-Z0-9\u4e00-\u9fa5)]/, '');
                    InquiryEdit_vue.demand.size = InquiryEdit_vue.demand.size.replace(/[^\d\.]/g, '');
                    InquiryEdit_vue.checkInfo();
                    InquiryEdit_vue.$watch(user.phone,'checkData');
                },
                //检验必填项
                checkInfo: function () {
                    // InquiryEdit_vue.checkPhone(InquiryEdit_vue.user.phone);

                    if (InquiryEdit_vue.checkPhone(InquiryEdit_vue.user.phone) && InquiryEdit_vue.user.phone && InquiryEdit_vue.user.nick && InquiryEdit_vue.demand.size) {
                        InquiryEdit_vue.clicked = false;
                    } else {
                        InquiryEdit_vue.clicked = true;
                    }
                },
                //正则校验手机号
                checkPhone: function (phone: string): boolean {
                    if (phone) {
                        if (!(/^1(3|4|5|7|8)\d{9}$/.test(phone))) {
                            InquiryEdit_vue.checkPhoneData = '请输入正确的手机号';
                            InquiryEdit_vue.showError = true;
                            return false;
                        } else {
                            InquiryEdit_vue.showError = false;
                            return true;
                        }
                    } else {
                        InquiryEdit_vue.checkPhoneData = '手机号不能为空';
                        return false;
                    }

                },
                openCitySel: function(){
                    tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
                        funs.loadPage(CitySelector);
                    })
                },
                //提交
                submit: function () {
                    console.log(InquiryEdit_vue.demand.type + "::" + InquiryEdit_vue.demand.location + "::" + InquiryEdit_vue.demand.color + "::" + InquiryEdit_vue.demand.span);
                    var phone = InquiryEdit_vue.user.phone;
                    var nick = InquiryEdit_vue.user.nick;
                    var type = parseInt(InquiryEdit_vue.demand.type);
                    var size = parseFloat(InquiryEdit_vue.demand.size);
                    var color = parseInt(InquiryEdit_vue.demand.color);
                    var location = parseInt(InquiryEdit_vue.demand.location);
                    var span = InquiryEdit_vue.demand.span;
                    var city = parseInt(InquiryEdit_vue.selected_city.city_id);
                    api.CreateDemand({ phone, nick, type, size, color, location, span, city }, function (data: any) {
                        // window.location.reload();
                        console.log(data.demand_id);

                        //未登录
                        if (model.globalData.user.user_type == 0) {
                            //send_sms_captcha==1表示已发送验证码
                            //无项目
                            if (data.demand_id && data.send_sms_captcha == 1) {
                                model.globalData.demand.id = data.demand_id;
                                model.globalData.demand.city_name = model.globalData.selected_city.city_name;
                                model.globalData.demand.user.nick = nick;
                                model.globalData.currIsSelfDemand = true;
                                model.globalData.current_demand = model.globalData.demand;
                                tomato.getVPresenter<chat>(funs.moduleToUrl('m/ChatRoom'), (chat) => {
                                    funs.loadPage(chat);
                                    chat.openTempChatRoom();
                                })
                            }
                            //demand_id==不存在，并且发送验证码，即报价失败，已有项目存在,弹出登录框
                            if ((!data.demand_id) && data.send_sms_captcha == 1) {
                                tomato.getVPresenter<LoginForm>(funs.moduleToUrl('m/user/LoginForm'), (LoginForm) => {
                                    LoginForm.login_init(phone);
                                    tdom.open(LoginForm, tdom.DialogTarget.Blank, {
                                        size: { width: '90%', height: tomato.DialogSize.Content },
                                        masked: true,
                                    })
                                })
                            }
                        } else {
                            //登录且无项目
                            if (data.demand_id) {
                                window.location.reload();
                            }else{
                                //登录且有项目
                                alert('您已有项目，不能再次创建');
                                setTimeout(function() {
                                    window.location.reload();
                                }, 1500);
                            }
                        }
                    }, function (error: tomato.PError) {
                        console.log(error);
                        if (error.name == '41004002') {
                            //此错误吗表示已存在招标中的项目
                            tomato.getVPresenter<LoginForm>(funs.moduleToUrl('m/user/LoginForm'), (LoginForm) => {
                                LoginForm.login_init(phone);
                                tdom.open(LoginForm, tdom.DialogTarget.Blank, {
                                    size: { width: '90%', height: tomato.DialogSize.Content },
                                    masked: true,
                                })
                            })
                        }
                        if(error.name == '41002011'){
                            tomato.getVPresenter<chat>(funs.moduleToUrl('m/ChatRoom'), (chat) => {
                                funs.loadPage(chat);
                                chat.openTempChatRoom();
                            })
                        }
                    });
                },
            },

        });

        this._watchEvent();

        this.addListener(tomato.VPresenterEvent.Installed, function() {
            $(view).find('.inquiry-total').removeClass('hide');
            $(view).find('.inquiry-list').addClass('hide');
        });

    }
    // //初始化询价
    public setProjeType(type) {
        this.InquiryEdit_vue.typePicked = type;
        this.InquiryEdit_vue.checkInfo();
        this.InquiryEdit_vue.chooseType();
    }
}
export = VPresenter;
