// views/m/project/Inquiry/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as Vue from 'vue';
import ChatRoom = require('views/m/ChatRoom');
import InquiryEdit = require('views/m/project/InquiryEdit');

class VPresenter extends project.VPresenter {
    private webim :any;

    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._watchEvent();
        var that = this;
        //展开收起
        //显示隐藏
        var demand_id = model.globalData.demand.id;
        //如果没有进行中的项目
        if (demand_id == '') {
            this.find('.no-demand').removeClass('hide');
            //城市
            var v_city = this.find('.location')[0];
            var city_info = model.globalData.selected_city;
            new Vue({
                el: v_city,
                data: {
                    city:city_info
                },
                methods: {
                    openCitySel: function (event) {
                        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
                            funs.loadPage(CitySelector);
                        })
                    }
                }
            });
            //点击切换屏幕类型
            var type_item = this.find('.panel-default')[0];
            var demand = model.globalData.demand;
            var type: any = new Vue({
                el: type_item,
                data: demand,
                // 在 `methods` 对象中定义方法
                methods: {
                    changeType: function (event) {
                        //修改选中类型
                        type.type_name = event.target.innerHTML;
                        type.type = $(event.target).attr("type_id");
                        var type_id = $(event.target).attr("type_id");
                        //收起选项
                        collapse.collapse('hide');
                        tomato.getVPresenter<InquiryEdit>(funs.moduleToUrl('m/project/InquiryEdit'), (InquiryEdit) => {
                            InquiryEdit.setProjeType(type_id);
                            funs.loadPage(InquiryEdit);
                        })
                    }
                }
            })
        } else {
            this.find('.has-demand').removeClass('hide');
            //登陆且有项目的用户，数据绑定
            var pro_inquiry = this.find('.has-demand')[0];
            var demd = model.globalData.demand;
            var unread = model.globalData.unread_num;
            var pro: any = new Vue({
                el: pro_inquiry,
                data: {
                    demand: demd,
                    unread: unread
                    },
                methods: {
                    openChatRoom: function (event) {
                        //将临时数据更改为自己的项目的数据
                        model.globalData.current_demand = model.globalData.demand;
                        model.globalData.currIsSelfDemand = true;
                        tomato.getVPresenter<ChatRoom>(funs.moduleToUrl('m/ChatRoom'), (ChatRoom) => {
                            //弹出详情页
                            funs.loadPage(ChatRoom);
                            ChatRoom.openChatRoom();
                            ChatRoom.setBudget();
                        })
                    }
                }
            });
            

            /*=====聊天室未读消息监听=====*/
            this.webim = {
            session: 'all',
            messages: { history: [], news: [] },
            update: function (msgs) {
                var private_unread_num = 0, group_unread_num = 0;;

                $.each(msgs, function (key, value) {
                    //其他会话的未读消息处理
                    if (value.news && value.unread > 0) {
                        if (model.globalData.demand.im_group.im_group_outer_id == '0') {
                            return
                        }
                        var gid = model.globalData.demand.im_group.im_group_outer_id;
                        if(key != gid) {
                            private_unread_num += value.unread;
                        }
                        group_unread_num = msgs[gid].unread;
                    }
                });
                //设置消息未读数量
                model.globalData.unread_num.group_unread = group_unread_num + '';
                model.globalData.unread_num.private_unread = private_unread_num + '';
            }
        };
            webimConfigs.addListener(this.webim);
        }
        var collapse = this.find('#Inquiry_collapse') as any;
        collapse.collapse('hide');

    }
    _evt_openInquiryEdit(){
        var type_id = this.find('.type-choosed').attr('type_id');
        tomato.getVPresenter<InquiryEdit>(funs.moduleToUrl('m/project/InquiryEdit'), (InquiryEdit) => {

            InquiryEdit.setProjeType(type_id);
            funs.loadPage(InquiryEdit);
        })
    }
}

export = VPresenter;
