// views/m/project/InquiryLayerBody/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import * as tdom from "@po-to/tomato-jquery";
import InquiryEdit = require('views/m/project/InquiryEdit');
import InquiryLayer = require('views/m/project/InquiryLayer');

class VPresenter extends project.VPresenter {
    /**模块 */
    private InquiryLayerBody_vue;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        var vueDom = this.find("#InquiryLayerBody_vue").get(0);
        var user = model.globalData.user;
        var demand = window['aaa'] = model.globalData.demand;
        var that = this;
        let InquiryLayerBody_vue: any = this.InquiryLayerBody_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                demand: demand,

            },
            methods: {
                chooseType: function (event) {
                    InquiryLayerBody_vue.demand.type = event.target.getAttribute('value');
                    InquiryLayerBody_vue.demand.type_name = event.target.innerHTML;
                    (that.parent as tomato.Dialog).close();
                    tomato.getVPresenter<InquiryLayer>(funs.moduleToUrl('m/project/InquiryLayer'), (InquiryLayer) => {
                       InquiryLayer.InquiryLayer_vue.typeClick=false;
                    })
                    //选择之后切到询价页去
                    tomato.getVPresenter<InquiryEdit>(funs.moduleToUrl('m/project/InquiryEdit'), (InquiryEdit) => {
                        InquiryEdit.setProjeType(demand.type);
                        console.log(demand.type);
                        funs.loadPage(InquiryEdit);
                    })
                }
            }
        });

        //this._watchEvent();
        //this._els = this._getElements();
    }
}
export = VPresenter;
